### DISCUSSION TITLE
Ordenação de propriedades no CSS

### DISCUSSION THREAD

**Discussion by eduardojmatos** [frontendbr/forum#484] [Discussão]
Sim, vai ser polêmica essa issue, mas a ideia é saber como vocês normalmente organizam as propriedades de CSS.

Estamos discutindo isso aqui na @getninjas e cada um tem uma opinião diferente. Existem duas formas de organização que estamos cogitando aqui:

## Alfabética

**PROS**
- A ideia é ser mais fácil de achar uma propriedade;
- É no-brainer, não precisa pensar muito antes de colocar as properties;

**CONTRAS**
- propriedades misturadas e longe do contexto das próximas;

## Por Tipo

**PROS**
- Faz pensar mais na UI, já que as definições de cada módulo de CSS fica próxima;

**CONTRAS**
- fica difícil de codar sem um SassLint;
- precisa entender dos módulos de CSS (?);

O CSS Tricks fez uma enquete sobre isso em 2012, vejam o resultado:

https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/

* * *

E vocês? Como fazem por aí?

**Comment by andrewsfg** [comment]
@eduardojmatos Eu sempre preferi por ordem alfabética, porque nela os debates são menores e é mais didático que todas outras, mas já vi várias abordagens por tipo também. Quando trabalhava na Folha de S.Paulo, adotamos o padrão alfabético e funcionou bem. O que talvez fique complicado é um left, top perdido um do outro.

**Comment by bernardodiasc** [comment]
Particularmente gosto muito do http://rhodesmill.org/brandon/2011/concentric-css/

Minha habilidade de CSS melhorou depois que compreendi os conceitos do concentric-css, mas é algo praticamente impossivel de se aplicar numa equipe onde o CSS é só uma mera bobagem pra deixar as coisas menos feias. Digo, é como a organização por tipo, apenas equipes interessadas em ter o CSS de qualidade vão se dar ao trabalho de seguir a regra.

**Comment by kvnol** [comment]
Eu costumo organizar por tipo.

**Posicionamento:**
position, top, left, right, bottom, z-index

**Box-model:**
width, height, display, margin, padding

**Tipografia:**
color, font-family (size, weight...), line-height

**Visual:**
background (image, color...), border 

Sigo a base do @mdo: http://codeguide.co/#css-declaration-order

**Comment by andrewsfg** [comment]
@kvnol interessante.

**Comment by alalbuquerque** [comment]
@kvnol sigo esse e ainda coloco as propriedade em ordem alfabética pra ficar separado.

**Comment by kvnol** [comment]
@andrewsfg Sim! Eu vejo esse método de escrita de CSS como uma alusão ao teclado QWERTY, parece desorganizado mas depois você percebe que é a "melhor forma" de escrever.

**Comment by felipefialho** [comment]
Boa issue @eduardojmatos, 

Já faz uns bons aninhos que ordeno por ordem alfabetica, inclusive, considero problema grave no lint. Acostumei e gosto muito.

**Comment by fdaciuk** [comment]
Sigo o padrão de organizar alfabeticamente também já faz alguns anos. No começo eu organizava por tipo, mas é como o @eduardojmatos postou: por ordem alfabética acaba se tornando muito automático.

Separando por tipo, ainda que você faça isso por muito tempo, uma hora ou outra vai acontecer de você duplicar a mesma propriedade, e perder alguns minutos/horas/dias/meses/anos pra achar o problema xD

E pensando em times, é muito mais fácil definir um padrão alfabético, que é comum a todo mundo desde a pré-escola, do que um padrão de TOC de alguém haha :stuck_out_tongue_winking_eye:

**Comment by kvnol** [comment]
@alalbuquerque Sim! Pode ser uma boa. Eu me acostumei com essa forma por alguns motivos, por exemplo: no posicionamento, eu começo sempre no top indo no sentido horário (top, right, bottom, left).

**Comment by kvnol** [comment]
Concordo com o @fdaciuk que é um TOC.

**Comment by lucasjs** [comment]
Nunca parei pra pensar nisso hahaha.
Sempre organizei por tipos. Acredito que é um bom formato . Não é difícil explicar, acho que o problema é é: as pessoas decorarem, ter que explicar tudo para novos colaboradores. Após ler sobre o que vocês falaram, vou ver se consigo propor para a galera aqui usar ordem alfabética. Você consegue explicar em segundos e as pessoas jamais vão esquecer.

**Comment by brunoravanhani** [comment]
Uso por tipo também, tipo o do @kvnol, agrupado e cada grupo separado por uma quebra de linha à mais,  mas fiquei com vontade de tentar organizar alfabeticamente agora.

**Comment by giulianovarriale** [comment]
Não sei se é pq eu agrupo por tipo, mas não me parece muito natural pensar em definir as propriedades do elemento em ordem alfabética. Normalmente, eu vou estilizando o componente por grupo. Primeiro tipografia, depois bloco, background, borda, posicionamento. Acho que, se eu trabalhasse em um projeto onde o padrão estabelecido fosse ordem alfabética, eu iria primeiro estilizar por grupos, depois rodar o lint e, depois dele me avisar que eu não coloquei em ordem alfabética, aí sim, alterar para ordem alfabética. Nunca tive problema de achar propriedades dentro de um seletor. Enfim, talvez seja só questão de hábito.

**Comment by XorAndre** [comment]
Eu não consigo trabalhar 100% organizado! Sei lá começar até começo, mas depois sai embaralhado :(

**Comment by dionarodrigues** [comment]
Desde que eu me entendo por desenvolvedor (rsss), sempre organizei por tipo. Houve uma época, numa empresa que trabalhei, que organizávamos por tipo e, em cada grupo de tipos, por ordem alfabética. No entanto, isso acabava virando bagunça quando se tinha muitas propriedades e muitas pessoas lidando com o mesmo código. Com o tempo acabei abrindo mão dá ordem alfabética por tipos.

Dias atrás estava avaliando organizar tudo por ordem alfabética, abrindo mão dos tipos, e essa issue tem sido de grande importância para que eu vejo como vocês organizam e tomar uma decisão mais assertiva. Vlw @eduardojmatos por levantar esse debate.

**Comment by eduardojmatos** [comment]
Eu sou mais favorável a adoção por tipo, e explico meus motivos aqui:

- Ordenação alfabética é mais fácil de já começar fazendo mesmo. Mas tudo é uma questão de adaptação. Ainda mais num quesito de conhecimento, já que pra ordernar por tipo você precisa conhecer os módulos de CSS. Acaba que virando um contra-argumento muito bom você usar por tipo, pra aprender de vez a separar bem os módulos 😸 ;

- Por tipo o aprendizado vem junto com adoção de uma ferramenta de lint, como o sasslint para o Sass. Dá pra configurar esse lint na ordenação, por exemplo;

- Ficar mais fácil de achar propriedades não consigo achar vantagem ser por ordem alfabética, pelo motivo de que, na maioria das vezes, não temos muitas propriedades num seletor apenas. É raro termos mais de 10;

No fim, provavelmente vamos tentar decidir por votação mesmo, porque ambos os lados tem bons pontos heheh

**Comment by leonardosouzafaria** [comment]
Estou adotando a organização por tipo ultimamente, além de eu ter mais afinidade com esse tipo de organização é uma forma de aprendizado para quem estar começando, para identificar o tipo de cada propriedade.

**Comment by danvitoriano** [comment]
E ae Edu. Eu acho as duas são legais. Sabe se tem algum plugin q faz isso depois q a gente escreve de qquer jeito?

**Comment by eduardojmatos** [comment]
@danvitoriano eu acho que o CSSComb não faz, e os lints tbm não. A única coisa é que vc pode impedir o build do Sass com o sasslint, por exemplo

**Comment by diessica** [comment]
@danvitoriano e @eduardojmatos:

O [stylefmt](https://github.com/morishitter/stylefmt) formata o CSS de acordo com as regras do stylelint. No caso, teria que usar o [stylelint-order](https://github.com/hudochenkov/stylelint-order) para configurar a regra da ordenação de propriedades.

**Comment by eduardojmatos** [comment]
@diessica 👏 Excelente! :)

**Comment by felipefialho** [comment]
@diessica 👏 👏 👏 👏 

PostCSS é embassado.. 😸

**Comment by bernardodiasc** [comment]
Penso assim como o @eduardojmatos, não se tratando como escovar paranoicamente os bits no CSS, mas do entendimento mais profundo de como as propriedades aplicam na interface e se afetam. Hoje em dia não uso nenhum tipo de ordenamento, mas carrego o aprendizado do concentric-css, minha forma de declarar CSS avançou consideravelmente, o tamanho do meu código reduziu tambem, já que não tem mais css macarronico nem propriedades inuteis ou redundantes no meio, consigo identificar bugs visuais muito mais facilmente e por ai vai.

**Comment by eduardojmatos** [comment]
Dando uma olhada na documentação do scss-lint, achei esses três tipos de ordenacão: concentric, recess e smaccs. Vale dar uma olhada na forma que estão divididos: https://github.com/brigade/scss-lint/tree/master/data/property-sort-orders

**Comment by akfzambrana** [comment]
Ja fiz em ordem alfabetica e hoje faço por tipo. Honestamente sigo 99% das coisas q tem no http://codeguide.co/ como o @kvnol comentou ai encima.

**Observação**: Apesar de eu "preferir" desse jeito, honestamente também, não ligo para tipo de ordenação, pois não influencia em absolutamente nada no quesito tempo de render, peso, performance, etc. logo é apenas uma preferencia, assim como gosto de "calça jeans classica". Acredito que não vale a pena ficar debatendo muito a questão se o time estiver dividido em opiniões (mesmo acreditando piamente na frase: 
> Every line of code should appear to be written by a single person, no matter the number of contributors.

Bjs!

**Comment by melissamoreira** [comment]
Ainda sou iniciante e sempre tento organizar por tipo, mas no fim acaba ficando tudo meio bagunçado.
Tópico muito interessante! Vou estudar as opções propostas acima.
Obrigada por levantar a questão :)

**Comment by kazzkiq** [comment]
Eu costumo fazer igual o @kvnol, propriedades CSS relacionadas à **posicionamento** e **dimensionamento** sempre no inicio do *bloco de declaração*, e propriedades relacionadas ao **visual** (mas que não afetem tanto a funcionalidade do elemento), sempre na parte de baixo do bloco.

Também tento manter as propriedades relacionadas a animações (`transform`, `transition`, `animation`) sempre no final dos blocos. Dessa forma o dev bate o olho em qualquer parte do CSS e já sabe mais ou menos em que região do bloco a propriedade vai estar.

**Comment by akfzambrana** [comment]
@melissamoreira uma forma q eu faço pra n ficar com "cara de bagunçado" é separar cada bloco com uma linha em branco, assim so de bater o olho ja percebo q é um bloco de tipo. Claro q tem uma curva de aprendizado enqto vc ainda não se acostuma a "separar", mas eu acho q vale muito a pena :smile:

**Comment by eubond** [comment]
@kazzkiq @kvnol 

Eu organizo por tipo, e acredito que seja a forma mais natural pra maioria das pessoas. Caso busquem uma ordem, aí sim, as pessoas passam a escolher entre as 2 opções citadas.

Todavia, fiquei com uma dúvida. 

Eu também prefiro fazer por tipo, porém, o grande problema disso pra mim é como definir a estrutura dos tipos?

Explico:

O @kazzkiq disse que coloca as propriedades de _"posicionamento e dimensionamento sempre no inicio do bloco de declaração"_. Eu, porém, prefiro começar com os de box-model e estilização. Que eu chamaria de construção.

Vocês utilizam essa ordem com base em algo ou só preferência? Caso não haja um motivo/regra, vejo que o principal problema de organizar por tipo, é que não seria um conhecimento compartilhado, visto que depende de como cada pessoa prefere organizar. Coisa que não acontece na ordem alfabética, que é um senso comum.

Valeu.

**Comment by kvnol** [comment]
@eubond Como disse, eu utilizo o Code-guide do @mdo como base para essa organização: http://codeguide.co/

**Comment by eubond** [comment]
@kvnol 

Poxa cara, vacilo meu. De qualquer forma, achei muito bacana a ideia e vou tentar implementar.

Valeu pela resposta.

**Comment by kazzkiq** [comment]
@eubond Eu particularmente sigo essa ordem por conta da relevância das propriedades.

Propriedades relacionadas ao **posicionamento** e **dimensionamento** afetam diretamente o funcionamento do elemento que irá utilizar a classe, e portanto, eu julgo serem as mais importantes, por isso ficam no topo.

As propriedades mais focadas no estilo visual (e não funcional) do elemento, podem ser modificadas sem afetar tanto como ele se comporta no layout de fato, por isso vem depois.

E as propriedades de animação, por serem responsáveis por algo que só vai acontecer com o elemento após ele estar **posicionado** e **visualmente estilizado**, ficam no final do bloco.

Não sei se é a melhor forma de fazer, mas pra mim faz muito sentido usar assim. 😄

**Comment by kvnol** [comment]
@eubond Que isso. Qualquer coisa, estamos aí.

**Comment by eubond** [comment]
@kazzkiq 

Faz sim cara e agora fez sentido pra mim tb. A minha organização, baseava-se na construção/identificação do elemento. Saca? Como se fosse a criação.

Eu gostei da referência que o @kvnol mandou e vou dar uma testada.

Valeu por responder mano, abriu a mente. =)

**Comment by waldemirflj** [comment]
Eu gosto do pattern "cascatinha". hehehe

````
.sub {
          top: 0;
          left: 0;
          width: 100%;
          z-index: -1;
          padding: em(10);
          overflow: hidden;
          position: absolute;
          background: #ffffff;
          @include transition(.1s);
          @include transform(translateY(-100%));
}
````

**Comment by bernardodiasc** [comment]
Só falta o padrão **acróstico** pra ordenar propriedades haha

**Comment by nandomoreirame** [comment]
Eu uso o padrão **O que vier na mente** eu vo colocando no código. :smile:

**Comment by eubond** [comment]
Dahora mano, freestyle. Haha

**Comment by wbotelhos** [comment]
Eita Batima, ficou pessoal hein?! Hahaha...

Bem, pela quantidade de respostas diversas e a curva de aprendizado em algo que será subjetivo, continuo com a opção que vem desde de lá do Jardim de Infância. Ordenação alfabética não impede o profissional de aprender Box Model e afins. A função de sort é largamente difundidas nos editores (Ex.: F5) e facilita achar propriedades duplicadas também.

No final, KISS e :shipit:

😗

**Comment by eduardojmatos** [comment]
@wbotelhos hahah ficou nada, você sabe que eu te amo 😘

No fim, decidimos por ordem alfabética no GetNinjas. A maioria concordou com o aspecto no-brainer de ser alfabético. 

Do mais, valeu pelos comentários galere! Discussão muito boa como sempre por aqui!

**Comment by wmsbill** [comment]
Pra mim eu só codo e pronto. Funcionou, tá igual? Tacalhe-pau.

Meu CSS é orientado a TOC. Mesma coisa para as minha variáveis JS

**Comment by eduardojmatos** [comment]
@wmsbill eu tbm, mas quando entra no time isso muda um pouco. Cada PR acaba ficando com um code style diferente e oxide review, que era pra ser de lógica e outras coisas acaba ficando só no detalhe de escrita

**Comment by didiraja** [comment]
Eu organizo elas por contexto com um olhar meio 3D em camadas. Começo por **background**, vou para os **tamanhos e bordas** _(border, margin, padding, width e height, nessa ordem)_, então começo a **tipografia** _(font-family, font-size, line-height)_ e depois vou completando com o que a classe exige.

Mas independente disso, se você for ordenar por contexto, acho válido você espaçar pelos contextos. Acho essa uma organização melhor que a alfabética, pq evita ficar corrigindo o css de propriedades q estão longe uma da outra.

```
.classe {
background-color:black;
margin: 0px 0px;
padding: 30px 20px;

font-family: 'Verdana', sans-serif;
font-size: 30px;
}
```

**Comment by thiamsantos** [comment]
Artigo relacionado: [Organize your CSS properties however you dang like](http://michael.blog/2017/03/30/organize-your-css-properties-however-you-dang-like/)

**Comment by felipefialho** [comment]
Só um adendo, essa discussão é realmente valida.

Um padrão na escrita das propriedades do CSS ajuda MUITO na mantenibilidade do código.

Peguei um projeto mais antigo onde eu ainda não definia padrões de propriedades (como disse, uso a-z atualmente) e mesmo eu, o próprio criador do código, demorei para me localizar, imaginem quem nunca teve contato com o projeto. 

Ao setar as propriedades por ordem alfabética, o seu olhar vai cair exatamente onde determinada propriedade deve estar.

**Comment by kvnol** [comment]
Não achei que diria isso, mas estou quase convencido em ordenar as propriedades em ordem alfabética. Uma das coisas que ainda não me convencem totalmente é deixar propriedades como `width` e `height` ou `top` e `bottom` longe umas das outras, pois são propriedades que se complementam e é, praticamente, automático (para mim) escrevê-las uma após a outra.

**Comment by felipefialho** [comment]
@knol É só questão de costume. No início parece não fazer sentido, depois é só vantagem.

**Comment by eduardojmatos** [comment]
@kvnol pode crer, sofro do mesmo mal. Não vejo sentido separa-los, massss vou descobrir como vai ser já que a galera decidiu usar assim lá no GetNinjas heheh

**Comment by kvnol** [comment]
@eduardojmatos sim, é o que realmente me impede de ir para a ordem alfabética. hahahaha

**Comment by bernardodiasc** [comment]
De um jeito ou de outro, o segredo é usar linter (hinter) até pegar total costume (aquele momento que o linter nao reclama mais pq fazemos certo de primeira).

**Comment by wmsbill** [comment]
Eu vou seguir a minha vidinha que já tá boa demais. Se um linter não fizer isso para mim, eu vou ficar de boa escrevendo CSS desordenado como se não houvesse amanhã :trollface: 

Até por que funciona.

**Comment by eduardojmatos** [comment]
@bernardodiasc sim! O linter vai me ajudar nesse processo hehehe

**Comment by eduardojmatos** [comment]
@wmsbill ah sim sim, funciona. O lance é quando você trampa num time e todos pensam de forma diferente pra organizar desde as propriedades até nomenclatura de classe, entre outras coisas. No fim isso ajuda a construir uma cultura do time.

E o linter vai ser o amigo fiel de todos nozes 🎉

**Comment by lucianobarauna** [comment]
Antes fazia por tipo e quando comecei a escrever o CSS por propriedades o ganho na manutenção ajudou bastante. No começo você acha coisa de doido e improdutivo mais com o tempo que vc vai ficando com isso na veia vc consegue sentir os benefícios.
Sigo o http://codeguide.co/ do mdo

**Comment by LucasBSousa** [comment]
@waldemirflj o *verdadeiro* "cascading" stylesheet

**Comment by mahenrique94** [comment]
Faz algum tempo que ordeno as props em ordem alfabética, para mim está sendo muito bom.

**Comment by claudiomborges** [comment]
Bom dia galera!
Como o assunto apareceu nas minhas notificações novamente, seria muito legal o @eduardojmatos nos passar suas impressões depois da adoção por ordem alfabética, as dores e alegrias durante o processo, e algo mais que ache pertinente. Valeu!
